### ISSUE TITLE
Open project from command line

### ISSUE THREAD

**Issue by julianorafael** [eveningkid/reacto#20]

**Comment by eveningkid** [comment]
I truly have no idea how to do this. Do you have any leads?
Also, is this necessary at the moment?

**Comment by julianorafael** [comment]
I found an article about that. I'll give it a try. Should the cli be called `reacto`? I feel like this would be good.

**Comment by eveningkid** [comment]
Sure, if you find **any details** about this and don't have time to implement it, share it there.
I can take care of it. :)
Let me know, so we can update this issue and make it available asap.

**Comment by julianorafael** [comment]
yeah, I did found out that you can use the app binary as a cli by calling `/Application/Reacto.app/Contents/MacOS/Reacto` and then handling the arguments on the `main.js` as explained here: https://discuss.atom.io/t/bundle-a-cli-app-inside-the-electron-app/40113/2
I'll be putting some time on this approach today.

**Comment by julianorafael** [comment]
https://github.com/atom/atom/blob/master/src/main-process/parse-command-line.js
atom deals with this using `yargs`

**Comment by eveningkid** [comment]
Really interesting approach. What are your thoughts on such a CLI helper for Reacto?

**Comment by julianorafael** [comment]
I'm not really sure it is needed if we are doing only one thing without any options. But if we want this to a be a more complete solution and have more options, `yargs` is indeed a nice solution. Short story: we might not need now, but we'll surely need it later.

**Comment by julianorafael** [comment]
Correction, right now it'd already be useful by parsing the options and showing the version when we type `reacto -v`.

**Comment by julianorafael** [comment]
The only problem with this approach is that I don't really know how to test this in development. Do you have any ideas on how to test this without making a package?

**Comment by eveningkid** [comment]
If we only get the current version of Reacto, I don't think that a CLI is currently needed. 
I guess you'll need to create a package even for testing. Simply set the `bin` attribute to where it should call the CLI from (quickly look this up this online, `bin` property of `package.json`). Your CLI will be just a node script I guess, right?

**Comment by julianorafael** [comment]
Ok, I agree that we do not need many options for now. The cli effectively is the own binary. We gotta handle the call from cli by parsing `process.argv` on `main.js` and checking if there's more arguments on it. If there is, than we check if the argument is a directory (I guess we should also open single files...) and then load the directory through the `filetree` function.
Is it only my machine that takes like 20+ minutes to pack the application?

**Comment by eveningkid** [comment]
I think your approach is correct and could work. I just have no idea how to actually do all the communication between a running reacto instance and what we get from the command line.

If you feel like spending time on this, please do some experiments and share your results here. I think that at this point, this is the only way to actually get answers. Maybe it won't work, maybe it will, haha.

And usually, I stop the build process when it says something like building "maps" or something. So usually, as soon as it updated the `.dmg` file, it's all good. Why asking? Do you need to pack it every single time for testing? :o

**Comment by simonhochrein** [comment]
This might be helpful, https://github.com/atom/atom/blob/d15e65a2c0101df280a60c9d99c03ff8153fffca/script/lib/install-application.js

**Comment by julianorafael** [comment]
Ok. Did some progress here. Now I'm parsing the command line and getting the desired path to open. The quest is to make it open the project. I believe I should send a message with `ipcMain` to the render process to trigger the `switchProject` action with the directory path. Does that makes sense? Where should this logic live on the application? Should I add a listener on the open project component to take care of this?

**Comment by eveningkid** [comment]
Somewhere inside [`public/main.js`](https://github.com/eveningkid/reacto/blob/master/public/main.js), when something is caught from the command line, simply call `mainWindow.webContents.send('switch-project', cwd)`.

Then, inside `src/editor/events`, create a file called `switch-project.js`, which simply dispatches the `switchProject` action (you can check other files in that folder, you'll quickly get the idea!).

Finally, add the listener to the `src/editor/events/index.js` file: import `switch-project`, add it to the `events` variable, and let the `EventsManager` listen to the `switch-project` event. Don't forget to pass it the `cwd` argument value.

I guess you have the whole recipe now! Thanks again for your help, I'm excited about this feature now!

**Comment by julianorafael** [comment]
Ok, I made it. It's opening a directory as project from the command line. Now I need to add a menu command to add reacto bin to PATH.
I can't test this on other OS (windows and linux). It might require some adjustments.

**Comment by eveningkid** [comment]
Amazing @julianorafael, please let us know when it's ready!!

**Comment by julianorafael** [comment]
I solved this by creating a symlink from the binary to `/usr/local/bin` on mac os. I added menu options to add and to delete this symlink and it only shows on `darwin` platforms for now. The only problem remainig is that the app remains attached to the console. I need to detach it, but I'm not sure how to do it yet.

**Comment by eveningkid** [comment]
@julianorafael could you please share your code so far so that could be easier to understand your issue at the moment?

**Comment by julianorafael** [comment]
Ok. I opened up a PR with the code. It appears that there has been some major changes on `main.js` and my code will need some changes. I couldn't find a way to do what I did on the `ready` event of the app with the `sweet-electron`. Also, as I said before, it opens from the terminal, but does not detach from it, which is quite annoying.

**Comment by eveningkid** [comment]
Thank you. I will add a `.ready((is, mainWindow) => ...)` method to `sweet-electron` tomorrow.
I guess it will be enough to adapt code from `main.js`. If not, please let me know and I'll add it as well. :)

For the menu, it should also be quite easy to fix the conflict, same for `events.js`.
Oh also, for the menu, I will add an instance of `electron-is` as `.menu(...)` first parameter so you'll be able to write `is.macOS()` inside `menu.js`.

For the terminal issue, is it different for Atom right now? If so, have you tried to find how they deal with it? 👀 

Thank you again for all your energy and time, I definitely appreciate it.

**Comment by julianorafael** [comment]
Your welcome sir, pleasure is mine. I'm a bit short on time, but I'm doing whatever I can. I'll find a way to deal with it and I'll refactor the code as soon as the method is available. The refactor is quite simple. The only thing is the ready event.
Thank you for the help and the attention, this is one of my first contributions with actual code to OSS projects. :)

**Comment by eveningkid** [comment]
Alright, I've just added the `.ready` method on `sweet()` instance! [You'll find an example here](https://github.com/eveningkid/sweet-electron/blob/master/examples/example.js#L7), but it is exactly as what I've explained in my previous message. :) (it should be fine, but if it's not, please let me know)

```js
.ready((is, mainWindow) => ...)
```

I am extremely glad that this is your first contribution to an open-source project! 

Keep me updated and keep it up!

**Comment by julianorafael** [comment]
![code](https://user-images.githubusercontent.com/1124448/38683557-240ab276-3e44-11e8-9798-2d4307d2da62.png)

**Comment by julianorafael** [comment]
I'm on version 1.0.7 of the sweet-electron

**Comment by julianorafael** [comment]
The function is there. But I guess that's not really the problem. I believe I need to create the window on this method, right? How do I access window configuration on the ready function to be able do that?

**Comment by julianorafael** [comment]
Nope, sweet-electron already takes care of that.

**Comment by julianorafael** [comment]
![image](https://user-images.githubusercontent.com/1124448/38685953-92173a14-3e49-11e8-8547-44c56c1459ff.png)
This is what I get when I try to open in development mode.

**Comment by julianorafael** [comment]
I can make a push to finish it :)

Don't worry about the delay, I really understand it. I'll make this push in the next few days, probably on the weekend.

I'll post it here if I find any issues along the way.

**Comment by eveningkid** [comment]
Thank you, thank you. Thank you for your time!

**Comment by julianorafael** [comment]
Status update: I'm working on it. Should submit in a couple of days.
